QEfficient/cloud/finetune.py

```diff
@@ -0,0 +1,204 @@
+# -----------------------------------------------------------------------------
+#
+# Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# -----------------------------------------------------------------------------
+
+import random
+import warnings
+
+import fire
+import numpy as np
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+import torch.optim as optim
+import torch.utils.data
+from peft import PeftModel, get_peft_model
+from torch.optim.lr_scheduler import StepLR
+
+from QEfficient.finetune.configs.training import train_config as TRAIN_CONFIG
+from QEfficient.finetune.utils.config_utils import (
+    generate_dataset_config,
+    generate_peft_config,
+    get_dataloader_kwargs,
+    update_config,
+)
+from QEfficient.finetune.utils.dataset_utils import (
+    get_custom_data_collator,
+    get_preprocessed_dataset,
+)
+from QEfficient.finetune.utils.train_utils import get_longest_seq_length, print_model_size, train
+
+try:
+    import torch_qaic  # noqa: F401
+except ImportError as e:
+    print(f"Warning: {e}. Moving ahead without these qaic modules.")
+
+
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+# Suppress all warnings
+warnings.filterwarnings("ignore")
+
+
+def main(**kwargs):
+    # update the configuration for the training process
+    train_config = TRAIN_CONFIG()
+    update_config(train_config, **kwargs)
+    device = train_config.device
+
+    # dist init
+    if train_config.enable_ddp:
+        # TODO: may have to init qccl backend, next try run with torchrun command
+        torch_device = torch.device(device)
+        assert torch_device.type != "cpu", "Host doesn't support single-node DDP"
+        assert (
+            torch_device.index is None
+        ), f"DDP requires specification of device type only, however provided device index as well: {torch_device}"
+        dist.init_process_group(backend=train_config.dist_backend)
+        # from here onward "qaic/cuda" will automatically map to "qaic:i/cuda:i", where i = process rank
+        getattr(torch, torch_device.type).set_device(dist.get_rank())
+
+    # Set the seeds for reproducibility
+    torch.manual_seed(train_config.seed)
+    random.seed(train_config.seed)
+    np.random.seed(train_config.seed)
+
+    # Load the pre-trained model and setup its configuration
+    # config = AutoConfig.from_pretrained(train_config.model_name)
+    model = AutoModelForCausalLM.from_pretrained(
```

Comment:
Question: 
Is it better to load model via QEffAutoModelForCausalLM (and in turn have all AI100 specific changes within) and then finetune, or load via default AutoModelForCausalLM, finetune, and then do AI100 specific changes for inference?

Reply:
We can do this in inference stage, as soon as we wrap in QEff model, structure changes due to AOT transforms.
We will do this in when we want to have e2e deployment and infer using the AOT stack.